include/memory_pool/shared_multi_heap.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021 Carlo Caione, <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_MEMORY_POOL_SHARED_MULTI_HEAP_H_
+#define ZEPHYR_INCLUDE_MEMORY_POOL_SHARED_MULTI_HEAP_H_
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Shared multi-heap memory pool interface
+ * @defgroup shared_multi_heap Shared multi-heap memory pool interface
+ * @ingroup multi_heap
+ * @{
+ *
+ * The shared multi-heap memory pool manager uses the multi-heap allocator to
+ * manage a set of reserved memory regions with different capabilities /
+ * attributes (cacheable, non-cacheable, etc...) defined in the DT.
+ *
+ * The user can request allocation from the shared pool specifying the
+ * capability / attribute of interest for the memory (cacheable / non-cacheable
+ * memory, etc...)
+ *
+ */
+
+/**
+ * @brief Memory region attributes / capabilities
+ *
+ * !!! This list needs to be kept in sync with shared-multi-heap-pool.yaml !!!
+ */
+enum smh_reg_attr {
+	SMH_REG_ATTR_CACHEABLE,
+	SMH_REG_ATTR_NON_CACHEABLE,
+	SMH_REG_ATTR_NUM,
```

Comment:
Is there a specific use case for these guys?  Seems like this is going to end up being very platform dependent.  Might want some kind of extension mechanism or something so every platform isn't having to come in and modify the global header.

Again, the use case I'm looking at are DSPs with three different memory regions of widely varying latencies (~12, 30 and 100+ cycles!), different power properties (one is low power always, but small, one has 64k banks that can be shut off programmatically), translation capability (one has a primitive global MMU, the others are direct mapped).  And each of those is actually mapped twice, once behind a per-CPU L1 cache and once with synchronized uncached access.  I despair of a general solution ever working for that, and even if it did no one would want to use that monster for simpler platforms.

tl;dr: This probably won't work for the audio DSP stuff, but that's a weird edge case.  Are we sure this is generic enough for "normal" apps?

Reply:
> Is there a specific use case for these guys?

Yup, there is. Having the having different memory regions with different cache-ability attributes is not uncommon, especially when you have multiple cores, each one behind a different cache for example.

But without going into more complex scenarios with multiple regions, I think this code is also handy also when for example you need to have one "no-cache" area for DMA where to take memory from. This is currently done in some drivers by leveraging the `CONFIG_NOCACHE_MEMORY` and some well-placed linker script symbols (see for example https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/dma/loop_transfer/src/test_dma_loop.c#L36 or https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/usb/device/usb_dc_mcux.c#L69). So in general is not uncommon to need heap allocation based on cacheability.

With this PR all the information is in the DT with a nice API on top to retrieve the needed memory heap.

> Might want some kind of extension mechanism or something so every platform isn't having to come in and modify the global header.

Yeah, I thought about that, but that would shift even more complexity and code from the pool manager to the user code and at least for now I'd like to keep this as simple as possible.

> Are we sure this is generic enough for "normal" apps?

I mean, it's probably not something that the wider audience will ever find useful but for sure this serves well for some more complex cases.